pyro/contrib/autoguide/__init__.py

```diff
@@ -561,6 +562,57 @@ def _loc_scale(self, *args, **kwargs):
         return loc, scale
 
 
+class AutoTransformedNormal(AutoContinuous):
+    """
+    This implementation of :class:`AutoContinuous` uses a Transformed
+    Low Rank MV Normal distribution via a IAF to construct a guide
+    over the entire latent space. The guide does not depend on the model's
+    ``*args, **kwargs``.
+
+    Usage::
+
+        guide = AutoTransformed(model, rank=10)
+        svi = SVI(model, guide, ...)
+
+    :param callable model: a generative model
+    :param int rank: the rank of the low-rank part of the covariance matrix
+    :param str prefix: a prefix that will be prefixed to all param internal sites
+    """
+    def __init__(self, model, hidden_dim=None, sigmoid_bias=2.0, prefix="auto", rank=1):
+        if not isinstance(rank, numbers.Number) or not rank > 0:
+            raise ValueError("Expected rank >= 0 but got {}".format(rank))
+        self.hidden_dim = hidden_dim
+        self.sigmoid_bias = sigmoid_bias
+        self.rank = rank
+        super(AutoTransformedNormal, self).__init__(model, prefix)
+
+    def sample_latent(self, *args, **kwargs):
+        """
+        Samples the (single) multivariate normal latent used in the auto guide.
+        """
+        loc = pyro.param("{}_loc".format(self.prefix),
+                         lambda: torch.zeros(self.latent_dim))
+        W_term = pyro.param("{}_W_term".format(self.prefix),
+                            lambda: torch.randn(self.rank, self.latent_dim) * (0.5 / self.rank) ** 0.5)
+        D_term = pyro.param("{}_D_term".format(self.prefix),
+                            lambda: torch.ones(self.latent_dim) * 0.5,
+                            constraint=constraints.positive)
+        hidden_dim = self.hidden_dim if self.hidden_dim is not None else self.latent_dim
+        iaf = dist.InverseAutoregressiveFlow(self.latent_dim, hidden_dim, sigmoid_bias=self.sigmoid_bias)
+        pyro.module("{}_iaf".format(self.prefix), iaf.module)
+        self.iaf_dist = dist.TransformedDistribution(dist.LowRankMultivariateNormal(loc, W_term, D_term), [iaf])
```

Comment:
I think we instead want to transform white noise via the iaf, then apply learnable diagonal noise:
```py
def sample_latent(self, *args, **kwargs):
    iaf = self._get_iaf()  # calls pyro.module internally
    iaf_dist = dist.TransformedDistribution(dist.Normal(0., 1.).expand([self.hidden_dim]), iaf)
    loc = pyro.sample("_{}_loc".format(self.prefix), iaf_dist.independent(1))
    scale = pyro.param("{}_scale".format(self.prefix), lambda: torch.ones(self.latent_dim))
    return pyro.sample("_{}_latent".format(self.prefix), dist.Normal(loc, scale).independent(1))

def _loc_scale(self, *args, **kwargs):
    iaf = self._get_iaf()  # calls pyro.module internally
    prototype = pyro.param("{}_scale".format(self.prefix))
    loc = iaf(prototype.new_zeros(self.hidden_dim))
    scale = None  # no analytical solution
    return loc, scale
```